pkg/atproto/labeler_firehose.go

```diff
@@ -23,30 +23,42 @@ import (
 
 func (atsync *ATProtoSynchronizer) StartLabelerFirehose(ctx context.Context, did string) error {
 	retryCount := 0
-	retryWindow := time.Now()
+	backoffDuration := time.Second
+	timeSinceStarted := time.Now()
 
 	for {
 		if ctx.Err() != nil {
 			return nil
 		}
 		err := atsync.StartLabelerFirehoseRetry(ctx, did)
 		if err != nil {
-			log.Error(ctx, "firehose error", "err", err)
-
-			// Check if we're within the 1-minute window
-			now := time.Now()
-			if now.Sub(retryWindow) > time.Minute {
-				// Reset the counter if more than a minute has passed
-				retryCount = 1
-				retryWindow = now
-			} else {
-				// Increment retry count if within the window
-				retryCount++
-				if retryCount >= 3 {
-					log.Error(ctx, "firehose failed 3 times within a minute, crashing", "err", err)
-					return fmt.Errorf("firehose failed 3 times within a minute: %w", err)
+			retryCount++
+			log.Error(ctx, "labeler firehose connection failed, retrying with backoff",
+				"err", err,
+				"retryCount", retryCount,
+				"backoffDuration", backoffDuration,
+				"labelerDID", did)
+
+			// Wait with exponential backoff before retrying
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-time.After(backoffDuration):
+				// Exponential backoff with max of 5 minutes
+				backoffDuration = backoffDuration * 2
```

Comment:
We've got a library I like for these already, see e.g. https://github.com/streamplace/streamplace/blob/c8583e1749c9eadaec5d5b961fbfef7ca7ce9cda/pkg/statedb/locks.go#L54-L71

5 minutes seems excessive; I can't imagine we're going to contribute to DDOSing anybody if it's a 30-second max (lots of places in the code already where 30 seconds is treated as an upper limit, such as "no segments found for 30 seconds, disconnecting ingest"